[clang][ASTMatcher] Fix execution order of hasOperands submatchers

nicovank opened the pull request:
`hasOperands` does not always execute matchers in the order they are written. This can cause issue in code using bindings when one operand matcher is relying on a binding set by the other. With this change, the first matcher present in the code is always executed first and any binding it sets are available to the second matcher.

Simple example with current version (1 match) and new version (2 matches):
```bash
> cat tmp.cpp
int a = 13;
int b = ((int) a) - a;
int c = a - ((int) a);

> clang-query tmp.cpp
clang-query> set traversal IgnoreUnlessSpelledInSource
clang-query> m binaryOperator(hasOperands(cStyleCastExpr(has(declRefExpr(hasDeclaration(valueDecl().bind("d"))))), declRefExpr(hasDeclaration(valueDecl(equalsBoundNode("d"))))))

Match #1:

tmp.cpp:1:1: note: "d" binds here
int a = 13;
^~~~~~~~~~
tmp.cpp:2:9: note: "root" binds here
int b = ((int)a) - a;
        ^~~~~~~~~~~~
1 match.

> ./build/bin/clang-query tmp.cpp
clang-query> set traversal IgnoreUnlessSpelledInSource
clang-query> m binaryOperator(hasOperands(cStyleCastExpr(has(declRefExpr(hasDeclaration(valueDecl().bind("d"))))), declRefExpr(hasDeclaration(valueDecl(equalsBoundNode("d"))))))

Match #1:

tmp.cpp:1:1: note: "d" binds here
    1 | int a = 13;
      | ^~~~~~~~~~
tmp.cpp:2:9: note: "root" binds here
    2 | int b = ((int)a) - a;
      |         ^~~~~~~~~~~~

Match #2:

tmp.cpp:1:1: note: "d" binds here
    1 | int a = 13;
      | ^~~~~~~~~~
tmp.cpp:3:9: note: "root" binds here
    3 | int c = a - ((int)a);
      |         ^~~~~~~~~~~~
2 matches.
```

If this should be documented or regression tested anywhere please let me know where.

AaronBallman reviewed (commented):
Oh, that's a good catch -- thank you for the fix!

Can you add a new test case in https://github.com/llvm/llvm-project/blob/main/clang/unittests/ASTMatchers/ASTMatchersTraversalTest.cpp so we don't regress this behavior accidentally in the future?

Also, please add a release note in https://github.com/llvm/llvm-project/blob/026d963cb004689477d2b5798cbba5ad41c25a70/clang/docs/ReleaseNotes.rst?plain=1#L347 so users know about the fix?

nicovank committed:
[clang][ASTMatcher] Fix execution order of hasOperands submatchers

The `hasOperands` matcher does not always execute matchers in the order they are
written. This can cause issues in code using bindings when one operand matcher
is relying on a binding set by the other. With this change, the first matcher
present in the code is always executed first and any binding it sets are
available to the second matcher.

nicovank commented:
Thanks a lot for the pointers! Added a test and a note 👍. Let me know if this is adequate.
If/when CI passes if everything looks in order you can press the merge button, I don't have commit access.

AaronBallman reviewed (approved):
LGTM!
